### PR TITLE
[dif, ibex] Add nmi api with unittests

### PR DESCRIPTION
**Pull request by engdoreis**
Add nmi api with unittests in the `dif_rv_core_ibex`
Signed-off-by: Douglas Reis <doreis@lowrisc.org>

### CODE REVIEW

File: sw/device/lib/dif/dif_rv_core_ibex.h [hunk 1]
```diff
@@ -170,6 +192,42 @@ dif_result_t dif_rv_core_ibex_clear_error_status(
...
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_rv_core_ibex_enable_nmi(const dif_rv_core_ibex_t *rv_core_ibex,
+                                         bool alert, bool wdog);
+
+/**
+ * Read the NMI enable status and current event state.
+ *
+ * @param rv_core_ibex Handle.
+ * @param nmi_state The nmi state.
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_rv_core_ibex.h:211]:
```suggestion
 * @param[out] nmi_state The nmi state.
```

File: sw/device/lib/dif/dif_rv_core_ibex.c [hunk 2]
```diff
@@ -205,3 +205,61 @@ dif_result_t dif_rv_core_ibex_clear_error_status(
 
   return kDifOk;
 }
+
+dif_result_t dif_rv_core_ibex_enable_nmi(const dif_rv_core_ibex_t *rv_core_ibex,
+                                         bool alert, bool wdog) {
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_rv_core_ibex.c:210]: maybe the API for this should be slightly different. feels a bit odd to return `kDifBadArg` when `alert` or `wdog` is false. Maybe this should be something like:

```suggestion
typdef enum dif_rv_core_ibex_nmi_source {
  kDifRvCoreIbexNmiSourceAlert = 0,
  kDifRvCoreIbexNmiSourceWdog = 0,
} dif_rv_core_ibex_nmi_source_t

dif_result_t dif_rv_core_ibex_enable_nmi(const dif_rv_core_ibex_t *rv_core_ibex,
                                         dif_rv_core_ibex_nmi_source_t nmi_source) {
```

wdyt?

File: sw/device/lib/dif/dif_rv_core_ibex.c [hunk 3]
```diff
@@ -205,3 +205,61 @@ dif_result_t dif_rv_core_ibex_clear_error_status(
...
+  nmi_state->alert_raised =
+      bitfield_bit32_read(reg, RV_CORE_IBEX_NMI_STATE_ALERT_BIT);
+  nmi_state->wdog_barked =
+      bitfield_bit32_read(reg, RV_CORE_IBEX_NMI_STATE_WDOG_BIT);
+
+  return kDifOk;
+}
+
+dif_result_t dif_rv_core_ibex_clear_nmi_state(
+    const dif_rv_core_ibex_t *rv_core_ibex, bool alert, bool wdog) {
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_rv_core_ibex.c:252]: same here, maybe a slightly different API?

File: sw/device/lib/dif/dif_rv_core_ibex.h (hunk 1 above)
Comment (engdoreis) [reply sw/device/lib/dif/dif_rv_core_ibex.h:211]: done

File: sw/device/lib/dif/dif_rv_core_ibex.c (hunk 2 above)
Comment (engdoreis) [reply sw/device/lib/dif/dif_rv_core_ibex.c:210]: You are right. So I changed the enable and clear API.

File: sw/device/lib/dif/dif_rv_core_ibex.c (hunk 3 above)
Comment (engdoreis) [reply sw/device/lib/dif/dif_rv_core_ibex.c:252]: You are right. So I changed the enable and clear API.

File: sw/device/lib/dif/dif_rv_core_ibex.c [hunk 4]
```diff
@@ -205,3 +205,69 @@ dif_result_t dif_rv_core_ibex_clear_error_status(
 
   return kDifOk;
 }
+
+dif_result_t dif_rv_core_ibex_enable_nmi(const dif_rv_core_ibex_t *rv_core_ibex,
+                                         dif_rv_core_ibex_nmi_source_t nmi) {
+  if (rv_core_ibex == NULL || (nmi & ~kDifRvCoreIbexNmiSourceAll) != 0) {
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_rv_core_ibex.c:211]: nit:

```suggestion
  if (rv_core_ibex == NULL || nmi & ~kDifRvCoreIbexNmiSourceAll) {
```

File: sw/device/lib/dif/dif_rv_core_ibex.c [hunk 5]
```diff
@@ -205,3 +205,69 @@ dif_result_t dif_rv_core_ibex_clear_error_status(
...
+      bitfield_bit32_read(reg, RV_CORE_IBEX_NMI_STATE_ALERT_BIT);
+  nmi_state->wdog_barked =
+      bitfield_bit32_read(reg, RV_CORE_IBEX_NMI_STATE_WDOG_BIT);
+
+  return kDifOk;
+}
+
+dif_result_t dif_rv_core_ibex_clear_nmi_state(
+    const dif_rv_core_ibex_t *rv_core_ibex, dif_rv_core_ibex_nmi_source_t nmi) {
+  if (rv_core_ibex == NULL || (nmi & ~kDifRvCoreIbexNmiSourceAll) != 0) {
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_rv_core_ibex.c:257]: same here:
```suggestion
  if (rv_core_ibex == NULL || nmi & ~kDifRvCoreIbexNmiSourceAll) {
```

**Review by timothytrippel — APPROVED** [review]
Thanks for fixes @engdoreis !

**Pull request merged**